tutorials/haskell/haskell-time-tutorial/tutorial.md

```diff
@@ -0,0 +1,481 @@
+
+We started by opening an [issue][issue] on GitHub. Quick parenthesis here,
+when you want to work on OSS I’d recommend to start opening an issue
+in the repo and discuss/brainstorm your idea with the maintainers,
+sometimes there are some good reasons why the feature you’re proposing
+have not been yet implemented and if you open a PR directly it could
+rejected, so save yourself a disappointment and open an issue first.
+Back to our issue, the maintainer of the library surprisingly proposed
+to implement these helper functions using pattern synonyms and I said
+surprisingly because I haven’t heard about them, so that’s why I’ll
+like to introduce you some basics about this Haskell feature that,
+actually, it has been around for some years now.
+
+
+## A little about pattern synonyms
+
+First of all I’d like to start with a caveat, this guide is not meant
+to explain pattern synonyms in depth, instead it’s meant to be a practical
+introduction to this concept.
+
+GHC documentation describes pattern synonyms like:
+
+> “Pattern synonyms enable giving names to parametrized pattern schemes.
+> They can also be thought of as abstract constructors that don’t have a
+> bearing on data representation.”
+
+And probably we’re in the same place as in the beginning so I’ve preferred
+to introduce pattern synonyms with a real world example. We’re going to
+borrow a pattern that exists in the time library. The [MonthOfYear][month-pattern]
+pattern allows us to define months of the year not by a number but by its name.
+
+The `MonthOfYear` data type is defined like this:
+
+```haskell
+type MonthOfYear = Int
+```
+
+As you can see, it is only an alias for `Int`. We now are able to define
+months like `Int`s where January will correspond to 1 and so on. But
+fortunately, since GHC 7.8 we are able to use pattern synonyms and we
+could, instead of using raw `Int`s, write this down in a pattern style
+and have a more idiomatic way of defining months.
+
+```haskell
+pattern January :: MonthOfYear
+pattern January = 1
+
+pattern February :: MonthOfYear
+pattern February = 2
+
+pattern March :: MonthOfYear
+pattern March = 3
+.
+.
+.
+
+-- | The twelve 'MonthOfYear' patterns form a @COMPLETE@ set.
+pattern December :: MonthOf
+pattern December = 12
+
+{-# COMPLETE January, February, March, April, May, June, July, August, September, October, November, December #-}Year
+```
+
+If you’re curious about why the `{-# COMPLETE #-}` pragma is being
+used, see the [COMPLETE pragma documentation][complete-pragma] for more
+information.
+
+Now if we need to pass a `MonthOfYear` as a parameter to a function
+we could write something like:
+
+```haskell
+foo November -- or foo 11 --Both will be equivalent
+```
+
+We’ll use this pattern in a function that basically takes a month and
+a day and yields if the corresponding Date is a holiday.
+
+```haskell
+dateToHoliday :: MonthOfYear -> DayOfMonth -> String
+dateToHoliday January  _   = "Happy new year!"
```

Comment:
Since you already introduced patterns before, I think we could use it here to deconstruct a `Day`:

```hs
dayToHoliday :: Day -> String
dayToHoliday (YearMonthDay _ January _) = ...
dayToHoliday (YearMonthDay _ November 1) = ...
dayToHoliday (YearMonthDay _ December 25) = ...
...
```